### ISSUE TITLE
Example error

### ISSUE THREAD

**Issue by bradygaster** [lostintangent/az-login#9]
If I clone the repo, `npm install`, and then f5 in vscode i get the following error. I'll debug a little more (the debugger actually doesn't attach). Also get same error when I run `npm test` at the command line. 

```
bradygaster@Bradys-MacBook-Pro:~/sources/az-login [master ≡]$ npm test

> az-login@0.2.2 test /Users/bradygaster/sources/az-login
> node ./examples

/Users/bradygaster/sources/az-login/examples/index.js:4
!async function () {
       ^^^^^^^^
SyntaxError: Unexpected token function
    at Object.exports.runInThisContext (vm.js:78:16)
    at Module._compile (module.js:543:28)
    at Object.Module._extensions..js (module.js:580:10)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.runMain (module.js:605:10)
    at run (bootstrap_node.js:420:7)
    at startup (bootstrap_node.js:139:9)
    at bootstrap_node.js:535:3
npm ERR! Test failed.  See above for more details.
```

**Comment by lostintangent** [comment]
Even though the az-login module itself only requires Node 6.9+ (because it uses ES6 features), the example Code currently uses async/await, which requires you to run it with Node.js 7.6+. We could easily convert it to use Promises directly, which would prevent this error.

**Comment by bradygaster** [comment]
So to clarify, the customer would need to have Node.js 7.6+ to be able to run the demo with the distro? Do we know the uptick in installations for that version and is there enough saturation to warrant it being a required dependency?

**Comment by lostintangent** [comment]
To clarify, this library is currently just an implementation detail of the Az.js and Serverless CLIs, and therefore, the example is largely meant for sanity testing this lib for myself, since it hasn't been directly consumed by any customers.

To consume the library only requires Node 6.9+ (which is the current LTS and recommended minimum). It's just the sample snippet that currently requires 7.6+, due to the use of async/await. I'll update it in a minute to just use Promises directly, which will remove the 7.6+ requirement.

**Comment by lostintangent** [comment]
@bradygaster The example is now updated and this issue should be resolved. Try pulling the latest from master and let me know if you run into any issues. I also added some additional context to the example code and `README` to provide a little more detail on the lib. 

As I mentioned, this module hasn't been directly consumed by anyone besides myself, so I'm definitely keen to get feedback on how to simplify the docs/API/etc. Thanks again!

**Comment by bradygaster** [comment]
@lostintangent AH got it. didn't mean to busy-work you. :) I just tried it and when it blew up I was concerned. I also think I have something odd going on when I try to use azjs from the command line - the login spins and spins and never actually completes. Could be an environmental thing. I'll update from this tweak you made to see if I can unblock myself. Thanks!

**Comment by lostintangent** [comment]
Definitely not busy work! It was pointless for me to couple the example code to recent Node versions, so this was a good reminder to fix that :)

Hmm, I haven't seen that hanging issue before with the login. Any Az.js (or Serverless) bugs with login would almost definitely be an issue with this library. Do you recall if this issue happened after completing a browser based login (as opposed to setting the env vars for a custom service principal)?

I'll look into this tonight since I'd definitely want to address any issues here ASAP.

**Comment by bradygaster** [comment]
I'm seeing the hanging issue with this at the moment, in the example. I had to update the package.info to add the resources arm client package - i figure this is something any experience node developer would know to do but i would **also** not call myself *an experienced node developer,* just a patient one. :)

If you don't mind I'll submit a PR with my change so this works out of the box. That said I'm still seeing the hang.

**Comment by bradygaster** [comment]
So when i run this in debug mode in vs code i never get to this line: 

```javascript
const { resourceGroups } = clientFactory(ResourceManagementClient);
```

I just hang after the call to `login().then()`.

**Comment by bradygaster** [comment]
last note on this and i'll stop bugging ya. :) this could be the issue - i'm logging in using an MSA (specifically @hotmail). Is this potentially the issue?

**Comment by bradygaster** [comment]
Okay I think that's the ticket. I need to login using an org account and not an MSA. Until we can fix MSA login, which of course, we should do right? RIGHT? :)

Also let me know how I can add support for multiple subscriptions and I'll dive in for fun times. 
<img width="657" alt="better-but-not-done-yet" src="https://user-images.githubusercontent.com/486363/27019966-aaf64814-4f6f-11e7-94d6-0f849d1d187d.png">

**Comment by lostintangent** [comment]
If your account has multiple subscriptions, `az-login`'s default behavior is to prompt you to select which one you'd like to use. That said, this logic is currently only used if the library is being ran from an interactive terminal (see [here](https://github.com/lostintangent/az-login/blob/master/index.js#L237)), which it can receive input from (i.e. have the user tell it which subscription to use).

When you are debugging from VS Code, there isn't an attached `stdin` to accept user input from, and therefore, `az-login` thinks that it can't prompt you. This is a "debugging bug" and not a product bug. If you simply run `npm test` to run the example (or `node examples`, etc.), instead of running the sample via the VS Code debugger, you shouldn't see the subscription issue above. We can definitely look into resolving this, however, debugging the sample app hasn't been a scenario that I've expected any consumer of this lib to need to do (unless they were helping me maintain it 👍 ).

I'm trying the MSA scenario right now to see if that's related to the hanging issue. You might be onto something with that, since I've had the bad habit of mostly testing with my MS org account. Stay tuned!

**Comment by lostintangent** [comment]
@bradygaster Yeah this issue is related to authenticating with an MSA. The login succeeds, but my code for generating a service principal (to perform subsequent logins "headlessly" for the dev) don't seem to work for MSAs. Looking into it now!

**Comment by bradygaster** [comment]
Thanks for the detailed info @lostintangent - let me know if I can re-test for you and if there are other items I can try out for ya. Keep up the great work!

**Comment by bradygaster** [comment]
Also verified that the cmd-line-only approach using npm test works with a multiple-subscription, non-MSA login. Excellent stuff.

**Comment by lostintangent** [comment]
Perfect! Thanks so much for verifying that. Hopefully I can figure out this MSA bug ASAP.

**Comment by lostintangent** [comment]
@bradygaster I've apparently hit a quota that is preventing me from testing Azure AD stuff anymore right now :) However, I pushed what I believe may be a fix to the `msa_fix` branch. If you get a chance, and want to try it out, let me know if it works. Otherwise, I'll pick up on it tomorrow, when my rate limit has hopefully been removed.

**Comment by fiveisprime** [comment]
Could this issue be related to the MSA auth issue?

https://github.com/serverless/serverless-azure-functions/issues/74

**Comment by lostintangent** [comment]
I'm pretty sure it is. I've got a potential fix, that I'm hoping to validate and publish this morning.

**Comment by lostintangent** [comment]
@bradygaster @fiveisprime Just sent our a PR that in my testing, resolves this issue. Would love some additional eyes and validation if you had the time. This is a pretty blocking issue, so I'd love to get it published, but I don't want to rush it out either :)

**Comment by lostintangent** [comment]
@bradygaster @fiveisprime I went ahead and merged this fix into master and would love some validation. In my testing, everything works as expected, both both Azure AD and MSA accounts.

**Comment by fiveisprime** [comment]
Sorry, I didn't have the bandwidth to get to this one today. I'll try for getting it tested tonight.

**Comment by bradygaster** [comment]
@lostintangent just saw this after a totally swamped day. i'll try it out now. with my fingers crossed!

**Comment by bradygaster** [comment]
@lostintangent i tried it and here's what i got. note, i deleted the `.azure` folder before i tried this to make sure I didn't have anything lying around. 

```json
{ Error: ENOENT: no such file or directory, open '/Users/bradygaster/.azure/azloginServicePrincipal.json'
    at Object.fs.openSync (fs.js:558:18)
    at Object.fs.writeFileSync (fs.js:1223:33)
    at Object.writeFileSync (/Users/bradygaster/sources/az-login/node_modules/jsonfile/index.js:115:13)
    at roleAssignments.create.then (/Users/bradygaster/sources/az-login/index.js:159:29)
    at process._tickCallback (internal/process/next_tick.js:103:7)
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/bradygaster/.azure/azloginServicePrincipal.json' }
```

**Comment by bradygaster** [comment]
**AH!!!** good news @lostintangent - I restored the .azure folder, then i deleted all the files in the folder rather than delete the *actual* folder. I re-tried the `npm test` and it worked!

**Comment by lostintangent** [comment]
Yeah the stack you just printed is a separate, but valid bug :) I'll submit another PR to address that issue, but I'm happy to see we're appearing to have resolved the MSA bug (another user confirmed the fix as well).

**Comment by bradygaster** [comment]
@lostintangent sweet, man! amazing work. this will help a ton of folks [like me] who continue to use MSAs or who are getting started. Huge value-add. Keep it up!

**Comment by bradygaster** [comment]
closing this issue as it has been resolved as well as spawning resolutions in a few other areas.
